### PR TITLE
Add A* algo

### PR DESCRIPTION
**Pull request by danil179**
Currently this A* doesn't support dynamic environment (including some turrets).

It's important to switch to our A* as our map is dynamic and has range of heuristics (including vision handling).

`GetPath` is currently unused as we get paths from the client.

### CODE REVIEW

**Review by lzardy — COMMENTED** [review]
Looks good, although the function IsAnythingBetween should probably be separated into two functions, one for pathing, one for vision. It would be for another PR.

File: GameServerLib/Content/NavGrid.cs [hunk 1]
```diff
@@ -54,22 +43,106 @@ public List<Vector2> GetPath(Vector2 from, Vector2 to)
...
+                var pathList = SmoothPath(new List<NavGridCell>(arr));
+                pathList.RemoveAt(0); // removes the first point
+                foreach (var navGridCell in pathList.ToArray())
+                {
+                    var cellPosition = TranslateFromNavGrid(new Vector<float>() { X = navGridCell.X, Y = navGridCell.Y });
+                    returnList.Add(new Vector2(cellPosition.X, cellPosition.Y));
+                }
+                return returnList;
+            }
+            //Logging.LoggerProvider.GetLogger().Debug("Found path: "+returnList.ToString());
```
Comment (Veykril) [inline GameServerLib/Content/NavGrid.cs]: Looks like you forgot a leftover debugging line

File: GameServerLib/Content/NavGrid.cs [hunk 2]
```diff
@@ -268,6 +341,23 @@ public int GetCellIndex(float x, float z)
...
             return -1;
         }
+        private List<NavGridCell> GetCellNeighbors(NavGridCell cell)
+        {
+            var x = cell.X;
+            var y = cell.Y;
+            List<NavGridCell> neighbors = new List<NavGridCell>();
+            for (int dirY = -1; dirY <= 1; dirY++)
+            {
+                for (int dirX = -1; dirX <= 1; dirX++)
```
Comment (FurkanS1821) [inline GameServerLib/Content/NavGrid.cs:350]: Damn this is way better than `List<NavGridCell> neighbors = { map[x-1,y], map[x, y-1], ... }` how did I never thought about this lmao

**Review by FurkanS1821 — APPROVED** [review]
I see nothing wrong.

File: GameServerLib/Content/NavGrid.cs (hunk 1 above)
Comment (danil179) [reply GameServerLib/Content/NavGrid.cs]: Removed.

**Pull request merged**